레스토랑 상세 페이지 구현하기

ljs614 opened the pull request:
안녕하세요. 6주차 라우터 구현 제출입니다.

우선은 Homepage에서 RestaurantsPage 이동만 구현했습니다.

react v6 에서는 Switch가 지원이 안 된다고 해서 Routes로 바꿨습니다.(App.test에서 에러가 나서 왜 그런지 찾아보니 Switch 문제였더군요...)

강의 보고 비슷하게 코드를 작성했지만 리뷰 부탁 드립니다. 
감사합니다.

ljs614 committed:
react-router-dom 설치

ljs614 committed:
RestaurantsPage생성

ljs614 committed:
App Homepage router 테스트 작성

ljs614 committed:
router 구현 및 HomePage 테스트 통과

ljs614 committed:
HomePage test 작성

ljs614 committed:
HomePage 테스트 통과

ljs614 committed:
App /restaurants router 테스트 작성

ljs614 committed:
App /restaurants router 테스트 통과

ljs614 committed:
HomPage path 대문자 소문자로 수정

moonkii commented on src/App.jsx:
```diff
@@ -1,32 +1,20 @@
+    <>
+      <h1>헤더</h1>
+      <Routes>
+        <Route exact path="/" element={<HomePage />} />
+        <Route path="/home" element={<HomePage />} />
```
`/home` 이 따로 필요한 이유가 있나요?

ljs614 replied:
Switch 오류 고칠 때 Route가 하나라서 뜨는 건가 하고 하나 더 추가했었는데 제거를 안 했네요.

moonkii commented on src/App.test.jsx:
```diff
@@ -1,34 +1,44 @@
-  expect(queryByText('한식')).not.toBeNull();
+describe('App', () => {
+  beforeEach(() => {
+    const dispatch = jest.fn();
```
`dispatch` 는 `beforeEach` 밖에서 선언해줘야 테스트할 때 따로 호출 테스트를 할 수 있을거에요.

오히려 여기서는 `mockClear()` 를 해주는게 필요할거에요!

moonkii commented on src/App.test.jsx:
```diff
@@ -1,34 +1,44 @@
+        <MemoryRouter initialEntries={['/']}>
+          <App />
+        </MemoryRouter>
+      ));
```
중복해서 작성해줘야하는 불편함이 있어서 path 만 매개변수로 받아주는 render 함수로 빼면 좋을 것 같아요.

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/148092196-fa0cab50-4fdd-47be-9443-99feda9c914e.png)

TDD 사이클에 맞춰 잘 하고 계신것 같아요 👍 

> 강의 보고 비슷하게 코드를 작성했지만 리뷰 부탁 드립니다.

혹시 강의와 비슷하게 코드를 베끼면 안된다고 생각하고 있으신가요?
만약 그렇다면 정반대에요!
코드숨에서는 베스트 프랙티스가 되는 코드의 패턴을 외워서 숙련도를 먼저 높이는걸 권장하고있어요.
응용하는건 숙련도가 높아지고나서 해도 충분하죠 ;)
오히려 저는 강의를 보고 그대로 따라하시는걸 추천드려요.
강의를 보지 않고 거의 외우듯이 강의와 똑같은 코드를 작성할 수 있으면 베스트에요.
빠르게 과제를 완성한 뒤 오히려 반복해서 만들어보시면 숙련도가 빠르게 올라갈거에요!

> react v6 에서는 Switch가 지원이 안 된다고 해서 Routes로 바꿨습니다.(App.test에서 에러가 나서 왜 그런지 찾아보니 Switch 문제였더군요...)

이 부분은 슬랙으로 다른 수강생 분들과 공유가능하실까요?
강의 촬영 이후에 업데이트 된 부분이라  아마 몰라서 문제를 겪으실 것 같은 분들이 많을 것 같아요!
코드숨에서는 함께 자라기를 추구해요 ;)

ljs614 committed:
리뷰 반영 수정

  App /home 제거
  App.test path를 매개변수로 받는 render 함수로 중복제거
  App.test dipatch는 beforeEach 밖으로, jest.clearAllMocks로 mock 초기화

ljs614 committed:
폴더 분류

ljs614 committed:
AboutPage 테스트 작성

ljs614 committed:
AboutPage 생성, 테스트 통과

ljs614 committed:
App 헤더 link to '/'

ljs614 committed:
App.test not found 테스트 작성

ljs614 committed:
NotFoundPage 생성, 테스트 통과

ljs614 committed:
webpack devServer 설정, error => NotFoundPage 로 이동됨

ljs614 commented:
Home 구현은 끝난 것 같네요. 내일 레스토랑 상세페이지 구현하겠습니다.
파일이 많아져서 컴포넌트를 폴더로 구분해 봤는데 이렇게 구분하면 될까요?
테스트는 테스트 폴더에 모아두는지 아니면 해당 컴포넌트랑 같이 넣어야 할지 모르겠네요.

ljs614 committed:
DetailPage 테스트 작성

ljs614 committed:
DetailPage 테스트 통과

ljs614 committed:
setDetail 테스트 작성

ljs614 committed:
setDetail 테스트 통과

ljs614 committed:
setRestaurantId 테스트 작성

ljs614 committed:
setRestaurantId 테스트 통과

ljs614 committed:
restaurantDetail api 테스트 작성, 액션 작성

ljs614 committed:
loadRestaurantId action테스트 작성

ljs614 committed:
loadRestaurantDetail 테스트 통과

ljs614 committed:
App /detail 테스트 작성

ljs614 committed:
App /detail 테스트 통과

ljs614 committed:
restaurant 클릭 이벤트 테스트 작성

ljs614 committed:
restaurant 클릭 이벤트 테스트 통과

ljs614 committed:
RestaurantDetailPage loading테스트 작성

moonkii commented:
> 파일이 많아져서 컴포넌트를 폴더로 구분해 봤는데 이렇게 구분하면 될까요?
테스트는 테스트 폴더에 모아두는지 아니면 해당 컴포넌트랑 같이 넣어야 할지 모르겠네요.

저는 pages, components 로 구분하는 편이고, 테스트는 따로 테스트 파일을 만들기 보다는 컴포넌트와 함께 관리해주는 편이에요.
이 부분은 정답이 있다기 보다는 스스로 인지하기 편한 구조대로 해보시면 좋을 것 같아요.
이렇게 저렇게 다양하게 하다보면 최적이다 생각이 드는 것들이 보일거에요 ;)

ljs614 committed:
detailPage loading  테스트 통과

ljs614 committed:
RestaurantDetailPage 구현 완료

ljs614 committed:
notfoundpage 수정

ljs614 committed:
Pages 폴더 밖으로 수정(리뷰 참고)

ljs614 commented:
우선 구현은 완료 했습니다.
상세페이지를 /detail로 구현했했는데
/restaurants/${id}로 다시 구현하겠습니다.

ljs614 committed:
restaurant detail 페이지 path 테스트 수정

ljs614 committed:
restaurants detail페이지 path 테스트 통과

ljs614 committed:
RestaurantsContainer setRestaurantId 제거

ljs614 committed:
loadRestaurantDetail action test 수정

  매개변수 받는 함수로 수정함

ljs614 committed:
loadRestaurantDetail action 테스트 통과

ljs614 committed:
reducer test restaurantId  관련 제거

  selectedRestaurantId 제거
  setRestaurantId 제거

ljs614 committed:
reducer restaurantId제거 테스트 통과

ljs614 committed:
restaurantDetailPage useParams 사용으로 변경

ljs614 committed:
restaurantsDetailContainer 테스트 작성

ljs614 committed:
restaurantDetailContainer 테스트 통과

ljs614 committed:
action setRestaurantId 제거

ljs614 committed:
Not Found 테스트 통과

ljs614 commented:
/:id 형식으로 상세페이지를 구현하긴 했는데
직접 /restaurants/1 을 입력하면 
/main.js 를 불러오는 과정에서 /restuarants/main.js 를 불러와서 404에러가 뜹니다.
아무리 검색을 해도 해결이 안 되네요😢

리뷰 부탁 드립니다.

ljs614 committed:
index.html /main.js 수정

ljs614 commented:
테스트 통과한 분들 참고하여 수정했습니다.
script의 src경로를 수정했어야 했군요.

moonkii commented:
> script의 src경로를 수정했어야 했군요.

해결책은 잘 찾으셨는데 원인은 왜 그런지 알아내셨나요?

moonkii commented on src/Components/RestaurantsContainer.test.jsx:
```diff
@@ -0,0 +1,33 @@
+
+describe('RestaurantsContainer', () => {
+  const dispatch = jest.fn();
+  useDispatch.mockImplementation(() => dispatch);
```
이 부분도 `beforeEach` 에 넣어주면 좋을 것 같아요.

ljs614 replied:
> > script의 src경로를 수정했어야 했군요.
> 
> 해결책은 잘 찾으셨는데 원인은 왜 그런지 알아내셨나요?

/ 이 없을 경우에는 현재 경로에서 /main.js를 찾지만 /가 있을 경우에는 절대 경로에서 찾기 때문에 그런 것 같네요.

moonkii commented on src/Pages/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,44 @@
+
+  context('with restaurant detail', () => {
+    it('render detail page', () => {
+      useSelector.mockImplementation((selector) => selector({
+        restaurantDetail: DETAIL,
```
반복되는 부분은 `beforeEach` 로 빼주시면 중복을 줄일 수 있고, 내부에 값을 다르게 세팅해주고 싶을 때는 given2 라이브러리를 사용해보시면 도움이 되실거에요 ;)

ljs614 replied:
given2 알아보겠습니다.

ljs614 commented:
>

ljs614 commented:
> > script의 src경로를 수정했어야 했군요.
> 
> 해결책은 잘 찾으셨는데 원인은 왜 그런지 알아내셨나요?

/가 없을 경우에는 상대 경로이기 때문에 현재 경로에서 /main.js를 찾지만 /가 있을 경우 절대 경로이기 때문에 최 상위에서 /main.js를 찾아서 정상 작동되는 것 같네요.
